File: Data Structures & Algorithms/two-integer-sum-ii/submission-4.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        for i in range(len(numbers)):
            diff = target - numbers[i]
            
            left = i + 1
            right = len(numbers) - 1
            while left <= right:
                mid = left + (right - left) // 2
                if numbers[mid] == diff:
                    return [i + 1, mid + 1]
                
                if numbers[mid] < diff:
                    left = mid + 1
                else:
                    right = mid - 1
            
        return []

File: Data Structures & Algorithms/two-integer-sum/submission-4.rs
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut seen: HashMap<i32, i32> = HashMap::new();

        for (i, &v) in nums.iter().enumerate() {
            let diff = target - v;
            if let Some(&index) = seen.get(&diff) {
                return vec![index, i as i32];
            }
            seen.insert(v, i as i32);
        }

        vec![]
    }
}
